elide-core/src/test/java/com/yahoo/elide/core/LifeCycleTest.java

```diff
@@ -395,6 +428,130 @@ public void testUpdateWithChangeSpec() {
         verify(onUpdateDeferredCallback, times(1)).execute(eq(book), isA(RequestScope.class), any());
         verify(onUpdateImmediateCallback, times(1)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitCallback, times(1)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(isA(Author.class), isA(RequestScope.class), any());
+
+        // verify no empty callbacks
+        verifyNoEmptyCallbacks();
+    }
+
+    @Test
+    public void testMultipleUpdateWithChangeSpec() {
+        Book book = mock(Book.class);
+
+        DataStoreTransaction tx = mock(DataStoreTransaction.class);
+
+        RequestScope scope = new RequestScope(null, null, tx, new User(1), null,
+                getElideSettings(null, dictionary, MOCK_AUDIT_LOGGER),
+                false);
+        PersistentResource resource = new PersistentResource(book, null, scope.getUUIDFor(book), scope);
+
+        verify(book, never()).onCreateBook(scope);
+        verify(book, never()).onDeleteBook(scope);
+        verify(book, never()).onUpdateGenre(any(RequestScope.class), any(ChangeSpec.class));
+        verify(book, never()).onUpdateGenre(any(RequestScope.class), isNull());
+        verify(book, never()).preRead(scope);
+        verify(book, never()).alwaysOnUpdate();
+        verify(onUpdateDeferredCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdateImmediateCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(book, never()).checkPermission(scope);
+
+        // Verify changeSpec is passed to hooks
+        resource.setValueChecked("genre", "new genre");
+        resource.setValueChecked("title", "new title");
+
+        verify(book, never()).onCreateBook(scope);
+        verify(book, never()).onDeleteBook(scope);
+        verify(book, times(1)).onUpdateGenre(any(RequestScope.class), any(ChangeSpec.class));
+        verify(book, never()).onUpdateGenre(any(RequestScope.class), isNull());
+        verify(book, times(1)).preRead(scope);
+        verify(book, never()).alwaysOnUpdate();
+        verify(onUpdateDeferredCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdateImmediateCallback, times(2)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(book, times(2)).checkPermission(scope);
+
+        scope.runQueuedPreSecurityTriggers();
+        verify(book, never()).onCreateBook(scope);
+        verify(book, never()).onDeleteBook(scope);
+        verify(book, times(1)).onUpdateGenre(any(RequestScope.class), any(ChangeSpec.class));
+        verify(book, times(1)).preRead(scope);
+        verify(book, never()).alwaysOnUpdate();
```

Comment:
Could we rename the alwaysOnUpdate function to alwaysOnUpdatePreCommit - so the test reads a bit better?

Reply:
Made all methods match the annotation.